packages/std/src/ibc/callbacks.rs

```diff
@@ -0,0 +1,173 @@
+    dest_callback: Option<IbcDstCallback>,
+}
+
+impl IbcCallbackData {
+    /// Use this if you want to execute callbacks on both the source and destination chain.
+    pub fn both(src_callback: IbcSrcCallback, dest_callback: IbcDstCallback) -> Self {
+        IbcCallbackData {
+            src_callback: Some(src_callback),
+            dest_callback: Some(dest_callback),
+        }
+    }
+
+    /// Use this if you want to execute callbacks on the source chain, but not the destination chain.
+    pub fn source(src_callback: IbcSrcCallback) -> Self {
+        IbcCallbackData {
+            src_callback: Some(src_callback),
+            dest_callback: None,
+        }
+    }
+
+    /// Use this if you want to execute callbacks on the destination chain, but not the source chain.
+    pub fn destination(dest_callback: IbcDstCallback) -> Self {
+        IbcCallbackData {
+            src_callback: None,
+            dest_callback: Some(dest_callback),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub struct IbcSrcCallback {
+    /// The source chain address that should receive the callback.
+    /// You probably want to put `env.contract.address` here.
+    pub address: Addr,
+    /// Optional gas limit for the callback (in Cosmos SDK gas units)
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub gas_limit: Option<Uint64>,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub struct IbcDstCallback {
+    /// The destination chain address that should receive the callback.
+    pub address: String,
+    /// Optional gas limit for the callback (in Cosmos SDK gas units)
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub gas_limit: Option<Uint64>,
+}
+
+/// The type of IBC source chain callback that is being called.
+///
+/// IBC source chain callbacks are needed for cases where your contract triggers the sending of an
+/// IBC packet through some other message (i.e. not through [`IbcMsg::SendPacket`]) and needs to
+/// know whether or not the packet was successfully received on the other chain.
+/// A prominent example is the [`IbcMsg::Transfer`] message. Without callbacks, you cannot know
+/// whether the transfer was successful or not.
+///
+/// Note that there are some prerequisites that need to be fulfilled to receive source chain callbacks:
+/// - The contract must implement the `ibc_source_chain_callback` entrypoint.
+/// - The module that sends the packet must be wrapped by an `IBCMiddleware`
+///   (i.e. the source chain needs to support callbacks for the message you are sending).
+/// - You have to add json-encoded [`IbcCallbackData`] to a specific field of the message.
+///   For `IbcMsg::Transfer`, this is the `memo` field.
+/// - The receiver of the callback must also be the sender of the message.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub enum IbcSourceChainCallbackMsg {
```

Comment:
`#[serde(rename_all = "snake_case")]` missing here?

Reply:
Yes, will add that.